[Farrel Dwireswara Salim] iP

sugiyem opened the pull request:
# Duke
> "Your mind is for having ideas, not holding them."  &ndash; David Allen ([source](https://dansilvestre.com/productivity-quotes/))

Duke frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,
1. download it from [here](https://github.com/sugiyem/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks.
4. let it manage the task for you :wink:

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)
---
If you're a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

sugiyem committed:
Level 1 - Implement Greet, Echo, Exit

sugiyem committed:
Level 2 - Implement Add, List

sugiyem committed:
Level 3 - Implement Mark, Unmark

sugiyem committed:
Level 4 - Implement ToDo, Event, Deadline

sugiyem committed:
A-TextUiTesting - Update input.txt and EXPECTED.txt

sugiyem committed:
Level 5 - Handle error & implement Exception

sugiyem committed:
Level 6 - Implement Delete

sugiyem committed:
Refactor code and use enums

sugiyem committed:
Update input.txt and EXPECTED.txt in test folder

sugiyem committed:
Add save functionality

sugiyem committed:
Add basic date handling

sugiyem committed:
Merge branch 'branch-level-8'

sugiyem committed:
Fix date object and storage incompatibility

sugiyem committed:
Add TaskList class

sugiyem committed:
Modify CommandHandler and Storage to use TaskList

sugiyem committed:
Add Ui class

sugiyem committed:
Modify CommandHandler to use Ui object

sugiyem committed:
Update the internal structure of Duke and Parser class

sugiyem committed:
Delete unused Printer class

sugiyem committed:
Change CommandException's name to DukeException

sugiyem committed:
Add duke packages

sugiyem committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

sugiyem committed:
Change access modifier of getStatusIcon in Task class to protected

sugiyem committed:
Add unit tests for parser, task, and ui package

sugiyem committed:
Refactor code related to Command class

sugiyem committed:
Add Javadoc

sugiyem committed:
Add find functionality

sugiyem committed:
Add functionality to find tasks by date

sugiyem committed:
Refactor code to comply with a coding standard

sugiyem committed:
Fix merge conflict

sugiyem committed:
Fix merge conflict

sugiyem committed:
Update text ui test

sugiyem committed:
Add checkstyle

sugiyem committed:
Refactor code to comply with coding style

sugiyem committed:
Add GUI to project

sugiyem committed:
Allow multiple searches for find and at command

sugiyem committed:
Fix conflicts after merging branch-A-Varargs

sugiyem committed:
Fix merge conficts after merging branch-Level-10

sugiyem committed:
Refactor code to pass the given checkstyle

sugiyem committed:
Add border radius and padding to dialog box

Elben85 commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,52 @@
+ * @author Farrel Dwireswara Salim
+ */
+public class FindCommand implements Command {
+    private final String[] keyWords;
```
shouldn't this be keywords since keyword is one word?

Elben85 commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,52 @@
+
+
+        this.keyWords = keyWord.split(",");
+    }
```
should keyWord be changed to keyword? I noticed this in several other places too.

Elben85 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,119 @@
+                        ? new Deadline(description, taskDate)
+                        : taskType.equals("E")
+                        ? new Event(description, taskDate)
+                        : new ToDo(description);
```
I like how you seperated the logical block

Elben85 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    /**
+     * Save the list of string to Duke's storage.
```
Should't this be saves?

Elben85 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,85 @@
+     * @param keyWords the strings of keyword.
+     * @return true if Task matches the keyword, false otherwise.
+     */
+    public boolean isContainKeyWord(String ... keyWords) {
```
Might it be better to change the variable name to containsKeyword?

Elben85 commented on src/test/java/duke/parser/ParserTest.java:
```diff
@@ -0,0 +1,42 @@
+    private final Parser sampleParser = new Parser();
+
+    @Test
+    public void parseText_inputEqualsBye_isListeningBecomeFalse() {
```
I like how you format the test methods correctly and each test methods are descriptive enough

Elben85 reviewed (commented):
Overall, I like the fact that you have adhered to almost all of the coding styling standard.

CJun1039 commented on src/main/java/duke/command/ToDoCommand.java:
```diff
@@ -0,0 +1,53 @@
+     * @param description the description of the ToDoCommand.
+     */
+    public ToDoCommand(String description) {
+        this.description = description;
+    }
+
+    /**
+     * Executes the ToDoCommand and returns the response pair.
+     *
+     * @param ui the Ui object to handle user interface.
+     * @param storage the storage used by the ToDoCommand.
+     * @param taskList the task list used by the ToDoCommand.
+     * @return the response pair.
+     * @throws DukeException If Duke fails to execute the ToDoCommand.
+     */
+    @Override
+    public Pair<Boolean, String> execute(Ui ui, Storage storage, TaskList taskList)
```
I like that you used noun for attributes and verbs for methods!

CJun1039 reviewed (commented):
Clean work and not many common coding standard violations. Good job! 👍 .

sugiyem committed:
Update build.gradle

sugiyem committed:
Add Assertions to TaskList methods

There are several methods in TaskList class that may not work when the input is invalid.
Using Assertions can help the Duke's user to detect any input mistake.
Let's add some Assertions to check the assumptions that should hold in TaskList methods.

sugiyem committed:
Refactor to improve code quality

The current code has some naming issues. There is also a large section of code grouped together that make it quite hard to read.
Improving code quality can make other developers' life easier.
Let's rename some variables and split large chunk of code to multiple groups of related statement.

sugiyem committed:
Use stream to construct AtCommand and FindCommand

sugiyem committed:
Merge pull request #1 from sugiyem/branch-A-Assertions

Add Assertions

sugiyem committed:
Merge branch 'master' into branch-A-CodeQuality

sugiyem committed:
Merge pull request #2 from sugiyem/branch-A-CodeQuality

Improve code quality

sugiyem committed:
Merge branch 'master' into branch-A-Streams

sugiyem committed:
Merge pull request #3 from sugiyem/branch-A-Streams

Add Streams

sugiyem committed:
Set up CI

sugiyem committed:
Add tags to Task object

sugiyem committed:
Modify Command and Storage to handle tags

sugiyem committed:
Add command to filter by tags

sugiyem committed:
Fix checkstyle

sugiyem committed:
Update unit tests

sugiyem committed:
Fix failed check

sugiyem committed:
Merge pull request #4 from sugiyem/branch-C-Tagging

Add Tagging System

sugiyem committed:
Update CI

sikai00 commented on src/main/java/duke/command/AllCommand.java:
```diff
@@ -0,0 +1,58 @@
+ * @author Farrel Dwireswara Salim
+ */
+public class AllCommand implements Command {
+    private final String[] tags;
```
Might be useful to create tags as its own class to prevent [Primitive Obsession](https://refactoring.guru/smells/primitive-obsession)!

sikai00 commented on src/main/java/duke/command/AllCommand.java:
```diff
@@ -0,0 +1,58 @@
+        // Filter out all empty tags
+        this.tags = Arrays.stream(chosenTags.split(","))
+                .map(String::trim)
+                .filter(text -> !text.isEmpty())
```
Use of double negative here, might be good to introduce isNotEmpty() [to remove double negatives](https://refactoring.com/catalog/removeDoubleNegative.html)

sikai00 commented on src/main/java/duke/command/AllCommand.java:
```diff
@@ -0,0 +1,58 @@
+     * @return the response pair.
+     */
+    @Override
+    public Pair<Boolean, String> execute(Ui ui, Storage storage, TaskList taskList) {
```
Nitpick, not sure what's the design decision behind but could use of Pairs to store results of command execution be confusing?

sikai00 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,88 @@
+        String mainCommand = commands.get(0);
+        String description = commands.size() > 1 ? commands.get(1) : "";
+
+        if (mainCommand.equals("bye")) {
```
Love how clean this is! But based on this context, maybe a switch statement here can help make it easier for readers to make sense of the code here.

sikai00 commented on src/main/java/duke/storage/Config.java:
```diff
@@ -0,0 +1,11 @@
+ * @author Farrel Dwireswara Salim
+ */
+public class Config {
+    public static final String DIRECTORY = "./data";
```
I like how you made these constants!

sikai00 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,131 @@
+                int dateSeparatorIndex = taskString.lastIndexOf("|");
+                int tagsIndex = taskString.indexOf("#");
+
+                String taskType = taskString.substring(0, 1);
```
There might be some potential to execute SLAP here, and make parsing of the string another function's job.

sikai00 replied:
I believe using SLAP here can also help make the code here much shorter and readable.

sikai00 reviewed (commented):
Overall your code is really clean! Especially with heavy use of streams to make it more 'functional style', making your code really readable. If there are any nitpicks it would only be the lengthy storage function that can be made shorter by applying SLAP. But otherwise, this is really great!

sugiyem committed:
Add a full stop at the end of the error message

sugiyem committed:
Add unit tests for duke.exception package

sugiyem committed:
Refactor Parser class using switch

sugiyem committed:
Add more test for Parser class

sugiyem committed:
Modify Task search to be case insensitive

sugiyem committed:
Add more unit tests for duke.task package

sugiyem committed:
Modify code to comply with coding standard

sugiyem committed:
Change font of text bar and submit button

sugiyem committed:
Make profile picture to be circle

sugiyem committed:
Differentiate Duke and User dialogs

sugiyem committed:
Merge pull request #6 from sugiyem/branch-A-MoreTesting

Add More Testing

sugiyem committed:
Merge pull request #7 from sugiyem/branch-A-BetterGui&Personality

Better GUI and More Personality

sugiyem committed:
Rename app to Duke Chatbot

sugiyem committed:
Add Ui.png to docs folder

sugiyem committed:
Update README.md

sugiyem committed:
Set jekyll theme

sugiyem committed:
Change the title of github pages

sugiyem committed:
Refactor code to handle SLAP violation

sugiyem committed:
Change fxml version to remove warning

sugiyem committed:
Change user's image from jpg to png

sugiyem committed:
Update build.gradle

sugiyem committed:
Disable window resize
